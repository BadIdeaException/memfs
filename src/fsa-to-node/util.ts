import {FsaToNodeConstants} from "./constants";;
import type {FsLocation} from "./types";

export const pathToLocation = (path: string): FsLocation => {
  if (path[0] === FsaToNodeConstants.Separator) path = path.slice(1);
  const lastSlashIndex = path.lastIndexOf(FsaToNodeConstants.Separator);
  if (lastSlashIndex === -1) return [[], path];
  const file = path.slice(lastSlashIndex + 1);
  const folder = path.slice(0, lastSlashIndex).split(FsaToNodeConstants.Separator);
  return [folder, file];
};
