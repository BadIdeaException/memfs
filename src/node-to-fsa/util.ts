import type {NodeFsaContext} from "./types";

/**
 * Creates a new {@link NodeFsaContext}.
 */
export const ctx = (partial: Partial<NodeFsaContext> = {}): NodeFsaContext => {
  return {
    ...partial,
    separator: '/',
  };
};

export const basename = (path: string, separator: string) => {
  const lastSlashIndex = path.lastIndexOf(separator);
  return lastSlashIndex === -1 ? path : path.slice(lastSlashIndex + 1);
};

const nameRegex = /^(\.{1,2})|(.*(\/|\\).*)$/;

export const assertName = (name: string, method: string, klass: string) => {
  const isInvalid = nameRegex.test(name);
  if (isInvalid) throw new TypeError(`Failed to execute '${method}' on '${klass}': Name is not allowed.`);
};
