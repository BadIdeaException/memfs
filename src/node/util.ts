import { ERRSTR } from './constants';
import * as errors from '../internal/errors';
import type { FsCallbackApi } from './types';
import type * as misc from './types/misc';

export function promisify(
  fs: FsCallbackApi,
  fn: string,
  getResult: (result: any) => any = input => input,
): (...args) => Promise<any> {
  return (...args) =>
    new Promise((resolve, reject) => {
      fs[fn].bind(fs)(...args, (error, result) => {
        if (error) return reject(error);
        return resolve(getResult(result));
      });
    });
}

export function validateCallback<T>(callback: T): misc.AssertCallback<T> {
  if (typeof callback !== 'function') throw TypeError(ERRSTR.CB);
  return callback as misc.AssertCallback<T>;
}

function _modeToNumber(mode: misc.TMode | undefined, def?): number | undefined {
  if (typeof mode === 'number') return mode;
  if (typeof mode === 'string') return parseInt(mode, 8);
  if (def) return modeToNumber(def);
  return undefined;
}

export function modeToNumber(mode: misc.TMode | undefined, def?): number {
  const result = _modeToNumber(mode, def);
  if (typeof result !== 'number' || isNaN(result)) throw new TypeError(ERRSTR.MODE_INT);
  return result;
}

export function nullCheck(path, callback?) {
  if (('' + path).indexOf('\u0000') !== -1) {
    const er = new Error('Path must be a string without null bytes');
    (er as any).code = 'ENOENT';
    if (typeof callback !== 'function') throw er;
    process.nextTick(callback, er);
    return false;
  }
  return true;
}

function getPathFromURLPosix(url): string {
  if (url.hostname !== '') {
    throw new errors.TypeError('ERR_INVALID_FILE_URL_HOST', process.platform);
  }
  const pathname = url.pathname;
  for (let n = 0; n < pathname.length; n++) {
    if (pathname[n] === '%') {
      const third = pathname.codePointAt(n + 2) | 0x20;
      if (pathname[n + 1] === '2' && third === 102) {
        throw new errors.TypeError('ERR_INVALID_FILE_URL_PATH', 'must not include encoded / characters');
      }
    }
  }
  return decodeURIComponent(pathname);
}

export function pathToFilename(path: misc.PathLike): string {
  if (typeof path !== 'string' && !Buffer.isBuffer(path)) {
    try {
      if (!(path instanceof require('url').URL)) throw new TypeError(ERRSTR.PATH_STR);
    } catch (err) {
      throw new TypeError(ERRSTR.PATH_STR);
    }

    path = getPathFromURLPosix(path);
  }

  const pathString = String(path);
  nullCheck(pathString);
  // return slash(pathString);
  return pathString;
}

const ENOENT = 'ENOENT';
const EBADF = 'EBADF';
const EINVAL = 'EINVAL';
const EPERM = 'EPERM';
const EPROTO = 'EPROTO';
const EEXIST = 'EEXIST';
const ENOTDIR = 'ENOTDIR';
const EMFILE = 'EMFILE';
const EACCES = 'EACCES';
const EISDIR = 'EISDIR';
const ENOTEMPTY = 'ENOTEMPTY';
const ENOSYS = 'ENOSYS';
const ERR_FS_EISDIR = 'ERR_FS_EISDIR';

function formatError(errorCode: string, func = '', path = '', path2 = '') {
  let pathFormatted = '';
  if (path) pathFormatted = ` '${path}'`;
  if (path2) pathFormatted += ` -> '${path2}'`;

  switch (errorCode) {
    case ENOENT:
      return `ENOENT: no such file or directory, ${func}${pathFormatted}`;
    case EBADF:
      return `EBADF: bad file descriptor, ${func}${pathFormatted}`;
    case EINVAL:
      return `EINVAL: invalid argument, ${func}${pathFormatted}`;
    case EPERM:
      return `EPERM: operation not permitted, ${func}${pathFormatted}`;
    case EPROTO:
      return `EPROTO: protocol error, ${func}${pathFormatted}`;
    case EEXIST:
      return `EEXIST: file already exists, ${func}${pathFormatted}`;
    case ENOTDIR:
      return `ENOTDIR: not a directory, ${func}${pathFormatted}`;
    case EISDIR:
      return `EISDIR: illegal operation on a directory, ${func}${pathFormatted}`;
    case EACCES:
      return `EACCES: permission denied, ${func}${pathFormatted}`;
    case ENOTEMPTY:
      return `ENOTEMPTY: directory not empty, ${func}${pathFormatted}`;
    case EMFILE:
      return `EMFILE: too many open files, ${func}${pathFormatted}`;
    case ENOSYS:
      return `ENOSYS: function not implemented, ${func}${pathFormatted}`;
    case ERR_FS_EISDIR:
      return `[ERR_FS_EISDIR]: Path is a directory: ${func} returned EISDIR (is a directory) ${path}`;
    default:
      return `${errorCode}: error occurred, ${func}${pathFormatted}`;
  }
}

export function createError(errorCode: string, func = '', path = '', path2 = '', Constructor = Error) {
  const error = new Constructor(formatError(errorCode, func, path, path2));
  (error as any).code = errorCode;

  if (path) {
    (error as any).path = path;
  }

  return error;
}

export function genRndStr6(): string {
  const str = (Math.random() + 1).toString(36).substring(2, 8);
  if (str.length === 6) return str;
  else return genRndStr6();
};
