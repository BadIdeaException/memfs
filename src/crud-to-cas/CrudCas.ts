import { hashToLocation } from './util';
import type { CasApi } from '../cas/types';
import type { CrudApi, CrudResourceInfo } from '../crud/types';

export interface CrudCasOptions {
  hash: (blob: Uint8Array) => Promise<string>;
}

const normalizeErrors = async <T>(code: () => Promise<T>): Promise<T> => {
  try {
    return await code();
  } catch (error) {
    if (error && typeof error === 'object') {
      switch (error.name) {
        case 'ResourceNotFound':
        case 'CollectionNotFound':
          throw new DOMException(error.message, 'BlobNotFound');
      }
    }
    throw error;
  }
};

export class CrudCas implements CasApi {
  constructor(protected readonly crud: CrudApi, protected readonly options: CrudCasOptions) {}

  public readonly put = async (blob: Uint8Array): Promise<string> => {
    const digest = await this.options.hash(blob);
    const [collection, resource] = hashToLocation(digest);
    await this.crud.put(collection, resource, blob);
    return digest;
  };

  public readonly get = async (hash: string): Promise<Uint8Array> => {
    const [collection, resource] = hashToLocation(hash);
    return await normalizeErrors(async () => {
      return await this.crud.get(collection, resource);
    });
  };

  public readonly del = async (hash: string, silent?: boolean): Promise<void> => {
    const [collection, resource] = hashToLocation(hash);
    await normalizeErrors(async () => {
      return await this.crud.del(collection, resource, silent);
    });
  };

  public readonly info = async (hash: string): Promise<CrudResourceInfo> => {
    const [collection, resource] = hashToLocation(hash);
    return await normalizeErrors(async () => {
      return await this.crud.info(collection, resource);
    });
  };
}
