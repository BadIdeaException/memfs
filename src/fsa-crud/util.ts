import {CrudType} from "../crud/types";
import {assertName} from "../node-to-fsa/util";

export const assertType = (type: CrudType, method: string, klass: string): void => {
  const length = type.length;
  for (let i = 0; i < length; i++) assertName(type[i], method, klass);
};
