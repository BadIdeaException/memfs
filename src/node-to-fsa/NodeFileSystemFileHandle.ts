import { NodeFileSystemHandle } from './NodeFileSystemHandle';
import { NodeFileSystemSyncAccessHandle } from './NodeFileSystemSyncAccessHandle';
import { basename, ctx as createCtx, newNotAllowedError } from './util';
import type { NodeFsaContext, NodeFsaFs } from './types';

export class NodeFileSystemFileHandle extends NodeFileSystemHandle {
  constructor(
    protected readonly fs: NodeFsaFs,
    public readonly __path: string,
    protected readonly ctx: Partial<NodeFsaContext> = createCtx(ctx),
  ) {
    super('file', basename(__path, ctx.separator!));
  }

  /**
   * Returns a {@link Promise} which resolves to a {@link File} object
   * representing the state on disk of the entry represented by the handle.
   * 
   * @see https://developer.mozilla.org/en-US/docs/Web/API/FileSystemFileHandle/getFile
   */
  public async getFile(): Promise<File> {
    try {
      const path = this.__path;
      const promises = this.fs.promises;
      const stats = await promises.stat(path);
      const data = await promises.readFile(path);
      const file = new File([data], this.name, {lastModified: stats.mtime.getTime()});
      return file;
    } catch (error) {
      if (error instanceof DOMException) throw error;
      if (error && typeof error === 'object') {
        switch (error.code) {
          case 'EPERM':
          case 'EACCES':
            throw newNotAllowedError();
        }
      }
      throw error;
    }
  }

  /**
   * @see https://developer.mozilla.org/en-US/docs/Web/API/FileSystemFileHandle/createSyncAccessHandle
   */
  public async createSyncAccessHandle(): Promise<NodeFileSystemSyncAccessHandle> {
    throw new Error('Not implemented');
  }

  /**
   * @see https://developer.mozilla.org/en-US/docs/Web/API/FileSystemFileHandle/createWritable
   */
  public async createWritable(
    { keepExistingData = false }: { keepExistingData?: boolean } = { keepExistingData: false },
  ): Promise<NodeFileSystemSyncAccessHandle> {
    throw new Error('Not implemented');
  }
}
