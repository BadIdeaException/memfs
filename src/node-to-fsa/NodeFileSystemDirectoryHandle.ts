import {NodeFileSystemHandle} from "./NodeFileSystemHandle";
import {assertName, basename, ctx as createCtx} from "./util";
import {NodeFileSystemFileHandle} from "./NodeFileSystemFileHandle";
import type {NodeFsaContext, NodeFsaFs} from "./types";

/**
 * @see https://developer.mozilla.org/en-US/docs/Web/API/FileSystemDirectoryHandle
 */
export class NodeFileSystemDirectoryHandle extends NodeFileSystemHandle {
  constructor (
    protected readonly fs: NodeFsaFs,
    protected readonly path: string,
    protected readonly ctx: Partial<NodeFsaContext> = createCtx(ctx),
  ) {
    super('directory', basename(path, ctx.separator!));
  }

  /**
   * Returns a new array iterator containing the keys for each item in
   * {@link NodeFileSystemDirectoryHandle} object.
   * 
   * @see https://developer.mozilla.org/en-US/docs/Web/API/FileSystemDirectoryHandle/keys
   */
  public async * keys(): AsyncIterableIterator<string> {
    const list = await this.fs.promises.readdir(this.path);
    for (const name of list) yield name;
  }

  /**
   * @see https://developer.mozilla.org/en-US/docs/Web/API/FileSystemDirectoryHandle/entries
   */
  public async * entries(): AsyncIterableIterator<[string, NodeFileSystemHandle]> {
    const {path, fs, ctx} = this;
    const list = await fs.promises.readdir(path, {withFileTypes: true});
    for (const dirent of list) {
      const name = dirent.name;
      const newPath = path + ctx.separator! + name;
      if (dirent.isDirectory()) yield [name, new NodeFileSystemDirectoryHandle(fs, newPath, ctx)];
      else if (dirent.isFile()) yield [name, new NodeFileSystemFileHandle(fs, name, ctx)];
    }
  }

  /**
   * Returns a new array iterator containing the values for each index in the
   * {@link FileSystemDirectoryHandle} object.
   * 
   * @see https://developer.mozilla.org/en-US/docs/Web/API/FileSystemDirectoryHandle/values
   */
  public async * values(): AsyncIterableIterator<NodeFileSystemHandle> {
    for await (const [, value] of this.entries()) yield value;
  }

  /**
   * @see https://developer.mozilla.org/en-US/docs/Web/API/FileSystemDirectoryHandle/getDirectoryHandle
   * @param name A string representing the {@link NodeFileSystemHandle} name of
   *        the subdirectory you wish to retrieve.
   * @param options An optional object containing options for the retrieved
   *        subdirectory.
   */
  public async getDirectoryHandle(name: string, options?: GetDirectoryHandleOptions): Promise<NodeFileSystemDirectoryHandle> {
    assertName(name, 'getDirectoryHandle', 'FileSystemDirectoryHandle');
    try {
      const filename = this.path + this.ctx.separator! + name;
      const stats = await this.fs.promises.stat(filename);
      if (!stats.isDirectory()) {
        throw new DOMException('The path supplied exists, but was not an entry of requested type.', 'TypeMismatchError');
      }
      return new NodeFileSystemDirectoryHandle(this.fs, filename, this.ctx);
    } catch (error) {
      if (error instanceof DOMException) throw error;
      if (error && typeof error === 'object') {
        switch (error.code) {
          case 'ENOENT': {
            throw new DOMException('A requested file or directory could not be found at the time an operation was processed.', 'NotFoundError');
          }
          case 'EPERM':
          case 'EACCES': {
            throw new DOMException('Permission not granted.', 'NotAllowedError');
          }
        }
      }
      throw error;
    }
  }

  /**
   * @see https://developer.mozilla.org/en-US/docs/Web/API/FileSystemDirectoryHandle/getFileHandle
   * @param name A string representing the {@link NodeFileSystemHandle} name of
   *        the file you wish to retrieve.
   * @param options An optional object containing options for the retrieved file.
   */
  public getFileHandle(name: string, options?: GetFileHandleOptions): Promise<NodeFileSystemDirectoryHandle> {
    throw new Error('Not implemented');
  }

  /**
   * @see https://developer.mozilla.org/en-US/docs/Web/API/FileSystemDirectoryHandle/removeEntry
   * @param name A string representing the {@link FileSystemHandle} name of the
   *        entry you wish to remove.
   * @param options An optional object containing options.
   */
  public removeEntry(name: string, options?: RemoveEntryOptions): Promise<void> {
    throw new Error('Not implemented');
  }

  /**
   * The `resolve()` method of the {@link FileSystemDirectoryHandle} interface
   * returns an {@link Array} of directory names from the parent handle to the specified
   * child entry, with the name of the child entry as the last array item.
   * 
   * @see https://developer.mozilla.org/en-US/docs/Web/API/FileSystemDirectoryHandle/resolve
   * @param possibleDescendant The {@link NodeFileSystemFileHandle} from which
   *        to return the relative path.
   */
  public resolve(possibleDescendant: NodeFileSystemHandle): Promise<string[] | null> {
    throw new Error('Not implemented');
  }
}

export interface GetDirectoryHandleOptions {
  /**
   * A boolean value, which defaults to `false`. When set to `true` if the directory
   * is not found, one with the specified name will be created and returned.
   */
  create?: boolean;
}

export interface GetFileHandleOptions {
  /**
   * A Boolean. Default `false`. When set to `true` if the file is not found,
   * one with the specified name will be created and returned.
   */
  create?: boolean;
}

export interface RemoveEntryOptions {
  /**
   * A boolean value, which defaults to `false`. When set to true entries will
   * be removed recursively.
   */
  recursive?: boolean;
}
