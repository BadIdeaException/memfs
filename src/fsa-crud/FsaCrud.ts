import type * as crud from '../crud/types';
import type * as fsa from '../fsa/types';
import {assertName} from '../node-to-fsa/util';
import {assertType} from './util';

export class FsaCrud implements crud.CrudApi {
  public constructor (protected readonly root: fsa.IFileSystemDirectoryHandle | Promise<fsa.IFileSystemDirectoryHandle>) {}

  protected async getDir(type: crud.CrudType): Promise<fsa.IFileSystemDirectoryHandle> {
    let dir = await this.root;
    for (const name of type)
      dir = await dir.getDirectoryHandle(name, {create: true});
    return dir;
  }

  public readonly put = async (type: crud.CrudType, id: string, data: Uint8Array, options?: crud.CrudPutOptions): Promise<void> => {
    assertType(type, 'put', 'crudfs');
    assertName(id, 'put', 'crudfs');
    const dir = await this.getDir(type);
    let file: fsa.IFileSystemFileHandle | undefined;
    switch (options?.throwIf) {
      case 'exists': {
        try {
          file = await dir.getFileHandle(id, {create: false});
          throw new DOMException('Resource already exists', 'ExistsError');
        } catch (e) {
          if (e.name !== 'NotFoundError') throw e;
          file = await dir.getFileHandle(id, {create: true});
        }
        break;
      }
      case 'missing': {
        try {
          file = await dir.getFileHandle(id, {create: false});
        } catch (e) {
          if (e.name === 'NotFoundError') throw new DOMException('Resource is missing', 'MissingError');
          throw e;
        }
        break;
      }
      default: {
        file = await dir.getFileHandle(id, {create: true});
      }
    }
    const writable = await file!.createWritable();
    await writable.write(data);
    await writable.close();
  };

  public readonly get = async (type: crud.CrudType, id: string): Promise<Uint8Array> => {
    throw new Error('Not implemented');
  };

  public readonly del = async (type: crud.CrudType, id: string): Promise<void> => {
    throw new Error('Not implemented');
  };

  public readonly info = async (type: crud.CrudType, id?: string): Promise<crud.CrudResourceInfo> => {
    throw new Error('Not implemented');
  };

  public readonly drop = async (type: crud.CrudType): Promise<void> => {
    throw new Error('Not implemented');
  };

  public readonly list = async (type: crud.CrudType): Promise<crud.CrudTypeEntry[]> => {
    throw new Error('Not implemented');
  };

  public readonly scan = async (type: crud.CrudType, cursor?: string | ''): Promise<crud.CrudScanResult> => {
    throw new Error('Not implemented');
  };
}
