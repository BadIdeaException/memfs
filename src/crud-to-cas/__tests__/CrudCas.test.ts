import { of } from 'thingies';
import { createHash } from 'crypto';
import { memfs } from '../..';
import { onlyOnNode20 } from '../../__tests__/util';
import { NodeFileSystemDirectoryHandle } from '../../node-to-fsa';
import { FsaCrud } from '../../fsa-to-crud/FsaCrud';
import { CrudCas } from '../CrudCas';

const hash = async (blob: Uint8Array): Promise<string> => {
  const shasum = createHash('sha1');
  shasum.update(blob);
  return shasum.digest('hex');
};

const setup = () => {
  const fs = memfs();
  const fsa = new NodeFileSystemDirectoryHandle(fs, '/', { mode: 'readwrite' });
  const crud = new FsaCrud(fsa);
  const cas = new CrudCas(crud, { hash });
  return { fs, fsa, crud, cas, snapshot: () => (<any>fs).__vol.toJSON() };
};

const b = (str: string) => {
  const buf = Buffer.from(str);
  return new Uint8Array(buf.buffer, buf.byteOffset, buf.byteLength);
};

onlyOnNode20('CrudCas', () => {
  describe('.put()', () => {
    test('can store a blob', async () => {
      const blob = b('hello world');
      const { cas, snapshot } = setup();
      const hash = await cas.put(blob);
      expect(hash).toBe('2aae6c35c94fcfb415dbe95f408b9ce91ee846ed');
      expect(snapshot()).toMatchSnapshot();
    });
  });

  describe('.get()', () => {
    test('can retrieve existing blob', async () => {
      const blob = b('hello world');
      const { cas } = setup();
      const hash = await cas.put(blob);
      const blob2 = await cas.get(hash);
      expect(blob2).toStrictEqual(blob);
    });

    test('throws if blob does not exist', async () => {
      const blob = b('hello world 2');
      const { cas } = setup();
      const hash = await cas.put(blob);
      const [, err] = await of(cas.get('2aae6c35c94fcfb415dbe95f408b9ce91ee846ed'));
      expect(err).toBeInstanceOf(DOMException);
      expect((<any>err).name).toBe('BlobNotFound');
    });
  });

  describe('.info()', () => {
    test('can retrieve existing blob info', async () => {
      const blob = b('hello world');
      const { cas } = setup();
      const hash = await cas.put(blob);
      const info = await cas.info(hash);
      expect(info.size).toBe(11);
    });

    test('throws if blob does not exist', async () => {
      const blob = b('hello world 2');
      const { cas } = setup();
      const hash = await cas.put(blob);
      const [, err] = await of(cas.info('2aae6c35c94fcfb415dbe95f408b9ce91ee846ed'));
      expect(err).toBeInstanceOf(DOMException);
      expect((<any>err).name).toBe('BlobNotFound');
    });
  });

  describe('.del()', () => {
    test('can delete an existing blob', async () => {
      const blob = b('hello world');
      const { cas } = setup();
      const hash = await cas.put(blob);
      const info = await cas.info(hash);
      await cas.del(hash);
      const [, err] = await of(cas.info(hash));
      expect(err).toBeInstanceOf(DOMException);
      expect((<any>err).name).toBe('BlobNotFound');
    });

    test('throws if blob does not exist', async () => {
      const blob = b('hello world 2');
      const { cas } = setup();
      const hash = await cas.put(blob);
      const [, err] = await of(cas.del('2aae6c35c94fcfb415dbe95f408b9ce91ee846ed'));
      expect(err).toBeInstanceOf(DOMException);
      expect((<any>err).name).toBe('BlobNotFound');
    });

    test('does not throw if "silent" flag is provided', async () => {
      const blob = b('hello world 2');
      const { cas } = setup();
      const hash = await cas.put(blob);
      await cas.del('2aae6c35c94fcfb415dbe95f408b9ce91ee846ed', true);
    });
  });
});
