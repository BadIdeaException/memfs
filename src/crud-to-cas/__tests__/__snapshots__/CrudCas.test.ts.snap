// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CrudCas .put() can store a blob 1`] = `
Object {
  "/ed/46/2aae6c35c94fcfb415dbe95f408b9ce91ee846ed": "hello world",
}
`;

exports[`CrudCas on FsaCrud .put() can store a blob 1`] = `
Object {
  "/ed/46/2aae6c35c94fcfb415dbe95f408b9ce91ee846ed": "hello world",
}
`;

exports[`CrudCas on NodeCrud .put() can store a blob 1`] = `
Object {
  "/ed/46/2aae6c35c94fcfb415dbe95f408b9ce91ee846ed": "hello world",
}
`;

exports[`CrudCas on NodeCrud at in sub-folder .put() can store a blob 1`] = `
Object {
  "/a/b/c/ed/46/2aae6c35c94fcfb415dbe95f408b9ce91ee846ed": "hello world",
}
`;

exports[`CrudCas on NodeCrud at root .put() can store a blob 1`] = `
Object {
  "/ed/46/2aae6c35c94fcfb415dbe95f408b9ce91ee846ed": "hello world",
}
`;
